specs/merkle-tree.md

```diff
@@ -0,0 +1,142 @@
+
+#### Input
+Accepts a [DeploymentData](https://github.com/sphinx-labs/sphinx/blob/591ff8a33578ecf61f184f4afc83a551397009f9/packages/contracts/src/merkle-tree.ts#L45C23-L45C23) object where the keys are canonical chain ids and the values are deployment data objects which contain all of the necessary info to assemble a SphinxMerkleTree. The Sphinx Merkle tree generation function should be agnostic to the source of the transaction data so it can be used to assemble a tree based on transactions generated from any scripting framework (i.e Foundry, Hardhat Ignition, some arbitrary future framework).
+
+#### Output
+Outputs a [SphinxMerkleTree](https://github.com/sphinx-labs/sphinx/blob/591ff8a33578ecf61f184f4afc83a551397009f9/packages/contracts/src/merkle-tree.ts#L115) object that follows the above specification and invariants.
+
+## Dependencies
+The Merkle tree generation logic makes calls to two battle-tested external libraries. We use `ethers` version 6.7.0 to handle ABI encoding tree leaf `data`, and we use `@openzeppelin/merkle-tree` version 1.0.5 to assemble the Merkle tree. We test that interactions with these libraries work properly in our [SphinxModuleProxy unit tests](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/packages/contracts/test/SphinxModuleProxy.t.sol) which cover that Merkle trees generated by this logic have data that is encoded properly and that they are executable on-chain. However, we do not test the internals of these libraries and instead rely on the assumption that they are robust and bug-free.
+
+## Assumptions
+The Merkle tree generation logic outputs trees which are valid and executable. However, there is no foolproof requirement that it be used. We make the asumption that relayers like the Sphinx DevOps platform will refuse to execute Merkle trees that are not valid according to the invariants defined in this document.
+
+Merkle trees generated with this logic are only valid to the extent that the data used to generate them is valid. We make the assumption that the input data satisfies the [on-chain conditions in the corresponding SphinxModuleProxy contract(s)](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/specs/sphinx-module-proxy.md#function-level-invariants).
+
+## Footnotes
+[^1]: Because the Sphinx Merkle tree contains leaves that are intended to be executed across multiple networks and not all leaves will be executed on all networks, we cannot rely on the position of the leaves within the tree to determine the leaf index. Instead, we use an explicit `index` field on each leaf.
+[^2]: Because Sphinx Merkle tree leaves contain explicit `index` fields, it is technically possible to assemble Merkle trees that have leaves which are ordered in a variety of ways (such as reverse order or simply random). We impose the ordering restriction to help make it easier to understand the content of the Merkle tree. It also benefits offchain relayer infrastructure.
```

Comment:
> it is technically possible to assemble Merkle trees that have leaves which are ordered in a variety of ways (such as reverse order or simply random)

An example of this would be great